stellarphot/photometry/photometry.py

```diff
@@ -1,138 +1,408 @@
-            bkgd, _, _ = sigma_clipped_stats(annulus_data * mask,
-                                             mask=bool_mask)
-            bkgd_pp.append(bkgd)
-        bkgd_pp = np.array(bkgd_pp)
-    else:
-        bkgd_pp = phot_table_1['aperture_sum'] / n_pix_ann
-
-    net_flux = phot_table['aperture_sum'] - (n_pix_ap * bkgd_pp)
-    phot_table['background_per_pixel'] = bkgd_pp
-    phot_table['net_flux'] = net_flux
-
-    # Return a columns with the aperture radius and
-    # the inner/outer annulus radii
-    phot_table['aperture_radius'] = \
-        np.ones(len(phot_table['aperture_sum'])) * aperture_radius
-    phot_table['inner_radius'] = \
-        np.ones(len(phot_table['aperture_sum'])) * inner_annulus
-    phot_table['outer_radius'] = \
-        np.ones(len(phot_table['aperture_sum'])) * outer_annulus
-
-    # Obtain RA/Dec coordinates and add them to table
-    try:
-        ra, dec = ccd_image.wcs.all_pix2world(coords[0], coords[1], 0)
-        phot_table['RA_center'] = ra
-        phot_table['Dec_center'] = dec
-    except AttributeError:
-        pass
+            med_pp.append(np.median(annulus_data))
+            std_pp.append(np.std(annulus_data))
+        photom['sky_per_pix_avg'] = photom['annulus_sum'] / photom['annulus_area']
+        photom['sky_per_pix_med'] = np.array(med_pp) * ccd_image.unit * u.pixel**-1
+        photom['sky_per_pix_std'] = np.array(std_pp) * ccd_image.unit * u.pixel**-1
+
+    # Compute counts using clipped stats on sky per pixel
+    photom['aperture_net_cnts'] = (photom['aperture_sum'].value -
+                                    (photom['aperture_area'].value *
+                                    photom['sky_per_pix_avg'].value))
+    photom['aperture_net_cnts'].unit = ccd_image.unit
+
+    # Fit the FWHM of the sources
+    print('single_image_photometry: fitting fwhm of all sources ...')
+    fwhm_x, fwhm_y = compute_fwhm(ccd_image, photom, fit=fwhm_by_fit)
+    print('                          ...DONE fitting fwhm of all sources')
+
+    # Deal with bad FWHM values
+    bad_fwhm = (fwhm_x<1) | (fwhm_y<1) # Set bad values to NaN now
+    fwhm_x[bad_fwhm] = np.nan
+    fwhm_y[bad_fwhm] = np.nan
+    photom['fwhm_x'] = fwhm_x * u.pixel
+    photom['fwhm_y'] = fwhm_y * u.pixel
+    photom['width'] = ((fwhm_x + fwhm_y) / 2) * u.pixel
 
-    # Obtain flux error and add column to return table
-    noise = np.sqrt(gain * net_flux + n_pix_ap * (1 + (n_pix_ap / n_pix_ann)) *
-                    (gain * (bkgd_pp + N_dark_pp) + (N_R**2)))
-    phot_table['aperture_sum_err'] = noise
+    # Flag sources with bad counts before computing noise
+    # This can happen, for example, when the object is faint
+    # and centroiding is bad.
+    bad_cnts = photom['aperture_net_cnts'].value < 0
+    all_bads = bad_cnts | bad_fwhm
+    photom['aperture_net_cnts'][all_bads] = np.nan
+
+    # Compute and save noise
+    noise = calculate_noise(gain=camera.gain.value,
```

Comment:
Seems like we should be passing in the `Camera` object instead of all of its attrtibutes.

Reply:
Also, can we pass in the `photom` table  and have `calculate_noise` extract what it needs from there?

Reply:
photom is not the final PhotometryData instance, so we can't do that, but I re-wrote it to use the Camera object (as well as re-writing all the tests).